byzcoin/tx_pipeline.go

```diff
@@ -327,13 +330,26 @@ func (p *txPipeline) processTxs(txChan <-chan ClientTransaction, initialState *t
 		p.wg.Add(1)
 		defer p.wg.Done()
 		intervalChan := getInterval()
+		var txHashes [][]byte
+	leaderLoop:
 		for {
 			select {
 			case tx, ok := <-txChan:
 				if !ok {
 					log.Lvl3("stopping txs processor")
 					return
 				}
+				for _, txHash := range txHashes {
+					if bytes.Compare(txHash, tx.Instructions.Hash()) == 0 {
+						log.Lvl2("Got a duplicate transaction, ignoring it")
+						continue leaderLoop
+					}
+				}
+				txHashes = append(txHashes, tx.Instructions.Hash())
+				if len(txHashes) > maxTxHashes {
```

Comment:
I think you're right we need limit this but I would prefer seeing that in the collection step so that we don't have multiple truncation in different places. In fact the collection already has the option to limit the number of transactions sent by the conodes so if a block is slow to be created we could ask for 0. What do you think ?

Reply:
Can you elaborate a bit more? I don't see what you're describing here. Where would you ask for `0`?

Reply:
My point is that instead of taking of the size of the array/map you're filling with hashes, the number of txs collected should be limited in the collection process so that you insure that no more than x transactions will be processed and then you don't need to clean the array/map (which make the use of the map interesting). You ask for `0` when you received too many transactions until the next block is created (when it's actually created, you start again to ask for >0 transactions).

Reply:
Hmm, I think I got most of it. The reason I'm using the `txHashes` is because transactions from previous blocks might also be double transactions. I had a first version that only made sure that all transactions from the current block are unique. But transactions might repeat also across blocks, so I added the `maxTxHashes`. But instead of counting the # of blocks where I make sure to have only unique hashes, I take the number of transactions that I want to be sure there are only unique hashes.

Does that make sense?